abdiff/core/calc_ab_metrics.py

```diff
@@ -0,0 +1,291 @@
+
+
+def calc_ab_metrics(
+    run_directory: str,
+    diffs_dataset: str,
+) -> dict:
+
+    os.makedirs(Path(run_directory) / "metrics", exist_ok=True)
+
+    # build field diffs dataframe
+    field_matrix_parquet = create_record_diff_matrix_parquet(run_directory, diffs_dataset)
+
+    # calculate metrics data from sparse matrix
+    metrics_data = calculate_metrics_data(field_matrix_parquet)
+
+    # update run data with metrics
+    update_or_create_run_json(
+        run_directory=run_directory, new_data={"metrics": metrics_data}
+    )
+
+    return metrics_data
+
+
+def create_record_diff_matrix_parquet(
+    run_directory: str,
+    diffs_dataset: str,
+    batch_size: int = 1_000,
+) -> str:
+    """Create a boolean sparse matrix of modified fields for all records.
+
+    This writes a single parquet file with rows for each record, and columns for each
+    TIMDEX field, and a value of integer 1 if that field has a diff and 0 if not.  This
+    provides a handy way to calculate aggregate metrics for a given field or source in
+    later steps.
+
+    This code does NOT use pyarrow, and momentarily creates a single dataframe in memory
+    for all rows.  This is safe given the nature of the dataframe: there may be 10m rows,
+    but there are only 20-30 columns, and either integer 1 or 0 as the value, resulting
+    in a very small dataset in memory despite a potentially high row count.
+    """
+    diffs_ds = ds.dataset(diffs_dataset)
+
+    batch_metrics_dfs = []
+    for i, batch in enumerate(
+        diffs_ds.to_batches(
+            batch_size=batch_size,
+            columns=["timdex_record_id", "source", "ab_diff"],
+        )
+    ):
+        t0 = time.time()
+        batch_df = batch.to_pandas()
+
+        # parse diff JSON to dictionary for batch
+        batch_df["ab_diff"] = batch_df["ab_diff"].apply(
+            lambda diff_json: json.loads(diff_json)
+        )
+
+        batch_metrics = []
+        for _, row in batch_df.iterrows():
+            record_metrics = {
+                "timdex_record_id": row["timdex_record_id"],
+                "source": row["source"],
+            }
+            diff_data = row["ab_diff"]
+            record_metrics.update(get_record_field_diff_bools_for_record(diff_data))
+            batch_metrics.append(record_metrics)
+
+        # build dataframe for batch
+        batch_metrics_df = pd.DataFrame(batch_metrics)
+        batch_metrics_dfs.append(batch_metrics_df)
+        logger.info(f"batch: {i+1}, elapsed: {time.time()-t0}")
+
+    # concatenate all dataframes into single dataframe for writing and replace None with 0
+    metrics_df = pd.concat(batch_metrics_dfs)
+    metrics_df = metrics_df.fillna(0)
+
+    # write parquet file via pandas writer
+    filepath = Path(run_directory) / "metrics" / "metrics.parquet"
+    metrics_df.to_parquet(filepath, index=False)
+    return str(filepath)
+
+
+def get_record_field_diff_bools_for_record(diff_data: dict) -> dict:
```

Comment:
This function is quite critical to this app: by parsing the JSON diff data, we are simply trying to understand if a given TIMDEX field has any changes.  This could be a new value, added or removed from the record, or even the order of the array slightly different.  The end results is a boolean `1` if that field had a difference.  These booleans values feed the sparse matrix above, which in turns feed the metrics aggregation.